dicts/dicts.gobra

```diff
@@ -0,0 +1,466 @@
+}
+
+// Keep all key-value pairs corresponding to the set of keys provided.
+ghost
+ensures forall x int :: {result[x]} (x in domain(d) && x in xs) ==>
+	(x in domain(result) && result[x] == d[x])
+ensures forall x int :: {x in domain(result)} x in domain(result) ==>
+	(x in domain(d) && x in xs)
+ensures domain(result) == xs intersection domain(d)
+opaque
+decreases xs
+pure func Restrict(d dict[int]int, xs set[int]) (result dict[int]int) {
+	return let ys := (xs intersection domain(d)) in
+		(sets.IsEmpty(ys) ? Empty() :
+			(let y := sets.Choose(ys) in
+			(let yr := sets.Remove(ys, y) in
+			(let _ := sets.IntersectLenUpper(xs, domain(d)) in
+			 Restrict(d, yr)[y = d[y]]))))
+}
+
+// Remove all key-value pairs corresponding to the set of keys provided.
+ghost
+decreases
+pure func RemoveKeys(d dict[int]int, xs set[int]) (result dict[int]int) {
+	return Restrict(d, domain(d) setminus xs)
+}
+
+// Remove a key-value pair. Returns unmodified dictionary if they key is not found.
+ghost
+ensures len(result) <= len(d)
+ensures k in domain(d) ==> len(result) == len(d) - 1
+ensures !(k in domain(d)) ==> len(result) == len(d)
+decreases
+pure func Remove(d dict[int]int, k int) (result dict[int]int) {
+	return let ys := (sets.Remove(domain(d), k)) in
+		let _ := util.Asserting(ys intersection domain(d) == ys) in
+		RemoveKeys(d, sets.SingletonSet(k))
+}
+
+// True iff k maps to the same value or is not in the domains of d1 and d2.
+ghost
+decreases
+pure func IsEqualOnKey(d1, d2 dict[int]int, k int) bool {
+	return !(k in domain(d1) || k in domain(d2)) ||
+		(k in domain(d1) && k in domain(d2) && d1[k] == d2[k])
+}
+
+// True iff if d1 is a subset of d2.
+ghost
+decreases
+pure func IsSubset(d1, d2 dict[int]int) bool {
+	return domain(d1) subset domain(d2) &&
+		forall k int :: {IsEqualOnKey(d1, d2, k)} {k in domain(d1)} (k in domain(d1)) ==>
+		IsEqualOnKey(d1, d2, k)
+}
+
+// Union of two dictionaries. Does not require disjoint domains: on the intersection,
+// values from the second dictionary are chosen.
+ghost
+ensures domain(result) == domain(d1) union domain(d2)
+ensures forall k int :: {result[k]} (k in domain(d2)) ==> result[k] == d2[k]
+ensures forall k int :: {result[k]}  (!(k in domain(d2)) && k in domain(d1)) ==>
+	result[k] == d1[k]
+opaque
+decreases domain(d1) union domain(d2)
+pure func Union(d1, d2 dict[int]int) (result dict[int]int) {
+	return let ks := domain(d1) union domain(d2) in
+		(sets.IsEmpty(ks) ? Empty() :
+			let k := sets.Choose(ks) in
+			let c1 := Remove(d1, k) in
+			let c2 := Remove(d2, k) in
+			let _ := sets.RemoveUnionLen(domain(d1), domain(d2), k) in
+			(k in domain(d2) ? Union(c1, c2)[k = d2[k]] : Union(c1, c2)[k = d1[k]]))
+}
+
+// Dictionaries are disjoint iff their domains are disjoint.
+ghost
+decreases
+pure func AreDisjoint(d1, d2 dict[int]int) bool {
+	return sets.AreDisjoint(domain(d1), domain(d2))
+}
+
+// The length of the union of two disjoint dictionaries is the sum of each of their lengths.
+ghost
```

Comment:
all lemmas should be opaque, please make sure that that is the case

Reply:
Originally, I didn't make them opaque as I expected that we would wait until we could make them closed. I will update gomaps and dicts; I will create an issue for the others so this PR doesn't get cluttered.